From 81a47153ad40280d024d96e036b26e88917cc2bf Mon Sep 17 00:00:00 2001
From: Om Narasimhan <omn@broadcom.com>
Date: Fri, 1 Mar 2013 10:17:39 -0800
Subject: [PATCH 708/762] usb2 : xhci/ehci class id bug workaround

In some revision of xlp2xx, class code on xhci pci device is wrong (i.e,
instead of xhci class code, ehci class code is read). As a workaround,
if the device is xlp2xx native controller, ehci initialization exits.

Based on Broadcom SDK 2.3.

Signed-off-by: Om Narasimhan <omn@broadcom.com>
Signed-off-by: Yanjiang Jin <yanjiang.jin@windriver.com>
---
 drivers/usb/host/ehci-pci.c |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/drivers/usb/host/ehci-pci.c b/drivers/usb/host/ehci-pci.c
index a2dbaf3..55caed7 100644
--- a/drivers/usb/host/ehci-pci.c
+++ b/drivers/usb/host/ehci-pci.c
@@ -24,6 +24,7 @@
 
 /* defined here to avoid adding to pci_ids.h for single instance use */
 #define PCI_DEVICE_ID_INTEL_CE4100_USB	0x2e70
+#include <asm/netlogic/hal/nlm_hal_xlp_dev.h>
 
 /*-------------------------------------------------------------------------*/
 
@@ -67,6 +68,10 @@ static int ehci_pci_setup(struct usb_hcd *hcd)
 		}
 		break;
        case PCI_VENDOR_ID_NETLOGIC:
+	       if(is_nlm_xlp2xx()){
+		       /* Should not proceed. No native EHCI */
+		       return -ENODEV;
+	       }
                if (pdev->device == 0x1007) {
 #if defined CONFIG_USB_EHCI_BIG_ENDIAN_MMIO
                         ehci->big_endian_mmio = 1;
-- 
1.7.0.4

